### ISSUE TITLE
GPU: multi-dimensional `irfft` lowering (since 0.10.0) is ~1.7× slower than the single cuFFT C2R op for Hermitian-symmetric inputs — request an opt-in for the direct path

### ISSUE THREAD

**Issue by hbahk** [jax-ml/jax#40143]
Since jax 0.10.0 (by #36383 on #29325), `jax/_src/lax/fft.py::_fft_lowering_gpu` (registered for `cuda` and
`rocm`) decomposes every multi-dimensional `IRFFT` into

```
transpose → complex IFFT over the outer axes → transpose → 1-D IRFFT on the last axis
```

so that GPU results match NumPy, which requires Hermitian symmetry only along the last
axis (cuFFT's multi-dimensional C2R assumes it on every transformed axis). That is the
right default. But for inputs that *are* Hermitian-symmetric on every axis — the spectrum
of a real array times a real centro-symmetric filter, i.e. the inner loop of every
FFT-convolution renderer — the two lowerings give identical results and the decomposition
is pure cost: two cuFFT launches and two transposes instead of one multi-dimensional C2R.
We measured it as a 1.7× slowdown of the transform and a 25–31% slowdown of a
real workload (a forced-photometry template renderer) between jax 0.9.0 and 0.10.0, on
both an H100 and an L40S, reproduced bit-for-bit equal outputs on both paths, and
bisected it to exactly this rule.

**Reproduction** ([jax_irfft_gpu_repro.py](https://github.com/user-attachments/files/31342376/jax_irfft_gpu_repro.py); jax + numpy only)

The script builds a Hermitian-symmetric input (`rfft2` of random real images × a
Gaussian PSF spectrum, layout `(49, 262, 80, 41)` complex64), times the default
`jnp.fft.irfft2(x, s=(80, 80))`, prints the HLO `fft` ops, then — for comparison only —
re-registers the generic lowering for the GPU platform to time the single-op C2R path on
the same bytes and checks the outputs.

```
$ CUDA_VISIBLE_DEVICES=1 python jax_irfft_gpu_repro.py          # L40S, card shared with another job
jax 0.11.1 | gpu NVIDIA L40S
input: (49, 262, 80, 41) complex64 (rfft2 layout of 49x262 real 80x80 images)
default lowering : HLO fft ops [('IFFT', '80'), ('IRFFT', '80')], transposes 2,    9.88 ms
direct C2R (opt) : HLO fft ops [('IRFFT', '80,80')], transposes 0,    4.59 ms
ratio default/direct = 2.16x ; max|diff|/max|y| = 2.3e-07

$ ... same script under jax 0.9.0
jax 0.9.0 | gpu NVIDIA L40S
default lowering : HLO fft ops [('IRFFT', '80,80')], transposes 0,    5.02 ms
no _fft_lowering_gpu in this jax (<0.10): default IS the single-op path
```

(The 2.3e-7 relative difference is float32 rounding between the two cuFFT
algorithms; on our production spectra — products of `rfft2` transforms — the two
paths were bit-identical.)

**Measurements** (same captured inputs, jitted, min of 7, `block_until_ready`; H100
80 GB, allocation-exclusive; jax wheels `jax[cuda12]==V`, python 3.12):

| jax | `irfft2`-heavy render, 49×262 stamps 80² | 49×16 grids 160² | HLO for one `irfft2` |
| --- | --- | --- | --- |
| 0.9.0 | 1.43 ms | 3.85 ms | 1 op: `fft_type=IRFFT, fft_length={80,80}` |
| 0.10.0 | 1.82 ms (+27%) | 5.05 ms (+31%) | `IFFT {80}` + transpose + `IRFFT {80}` (+2 transposes) |
| 0.11.1 | same as 0.10.0 |     | same |

(0.6.0–0.8.0 behave like 0.9.0; 0.5.3 within 5% of it.) Isolated transform on an L40S,
jax 0.11.1 vs 0.9.0: `(49,262,80,41)→(80,80)` 8.2 ms vs 4.8 ms (1.7×); `(49,16,160,81)→
(160,160)` 3.8 vs 0.9 ms; `max|Δ|/max|y| = 0` between the two lowerings on these inputs.
`jax.lax.fft(x, FftType.IRFFT, (80, 80))` lowers through the same rule, so there is no
user-level way to request the direct C2R.

**Proposal**

1. An opt-in for callers who guarantee full Hermitian symmetry, e.g. a config flag
  `jax_gpu_irfft_assume_hermitian` (default `False`, NumPy semantics unchanged) checked at
  the top of `_fft_lowering_gpu`, or an argument threaded through `lax.fft` /
  `jnp.fft.irfftn` — one branch plus tests (equality on symmetric inputs under both
  paths; NumPy agreement on a non-symmetric input with the flag off). Happy to send the PR.
2. Longer term, an exact lowering without the transposes (outer-axis C2C via strided
  cuFFT plans on non-trailing dims, which needs XLA's FFT thunk to accept them), or a
  single full C2C IFFT of the Hermitian-extended last axis plus the real part (one launch,
  no transposes, ~2× the FFT traffic) — worth measuring.

**Environment**: NVIDIA H100 80GB HBM3 (CUDA 12.9 wheels) and L40S (CUDA 12.8 wheels),
Linux x86_64, python 3.12/3.13, `jax[cuda12]` 0.5.3 … 0.11.1 from PyPI.
